content/v2.0/reference/storage-engine.md

```diff
@@ -0,0 +1,174 @@
+---
+title: InfluxDB storage engine
+description: >
+  An overview of the InfluxDB storage engine architecture.
+weight: 7
+menu:
+  v2_0_ref:
+    name: Storage engine
+v2.0/tags: [storage engine, internals, platform]
+---
+
+The InfluxDB storage engine ensures the following three things:
+
+- Data is safely written to disk
+- Queried data is returned complete and correct
+- Data is accurate (first) and performant (second)
+
+This document details the internal workings of the storage engine.
+This information is presented both as a reference and to aid those looking to maximize performance.
+
+Major topics include:
+
+* [Write Ahead Log (WAL)](#write-ahead-log-wal)
+* [Time-Structed Merge Tree (TSM)](#time-structured-merge-tree-tsm)
+* [Time Series Index (TSI)](#time-series-index-tsi)
+
+{{% note %}}
+##### At a glance: changes to the InfluxDB storage engine from 1.x to 2.0
+- The InfluxDB 2.0 storage engine no longer partitions data into shards by time.
+- **Buckets** replace databases and retention policies.
+- Only TSI is used. There is no more in-memory index.
+- The configuration interface and options have changed. Configuration options are not currently exposed, but will be.
+
+Read about the [v1 storage engine](https://docs.influxdata.com/influxdb/v1.7/concepts/storage_engine).
+{{% /note %}}
+
+## Writing data: from API to disk
+
+The storage engine handles data from the point an API request is received through writing it to the physical disk.
+Data is written to InfluxDB using [Line protocol](/) sent via HTTP POST request to the `/write` endpoint.
+Batches of points are sent to InfluxDB.
+Those batches are compressed and written to a WAL for immediate durability.
+The points are also written to an in-memory cache so that newly written points are immediately queryable.
+The cache is periodically written to disk as TSM files.
+As TSM files accumulate, they are combined and compacted into higher level TSM files.
+
+Points can be sent individually; however, for efficiency, most applications send points in batches.
+A typical batch ranges in size from hundreds to thousands of points.
+Points in a POST body can be from an arbitrary number of series.
+Points in a batch do not have to be from the same measurement or tagset.
+
+## Write Ahead Log (WAL)
+
+To ensure durability, we use a Write Ahead Log (WAL).
+<!-- The WAL is a write-optimized storage format that allows for writes to be durable, but not easily queryable -->
+<!-- On the write side, -->
+WAL is a data structure and algorithm that is super simple and powerful.
+It ensures that written data does not disappear when storage engine restarts.
+When a client sends a write request, the following occurs:
+
+1. Write request is appended to the end of the WAL file.
+2. fsync() the data to the file.
+3. Update the in-memory database.
+   <!-- 3. Update the in-memory CACHE? -->
+4. Return success to caller.
+
+fsync() is a system call, so it has a kernel context switch which costs something.
```

Comment:
maybe start w line 68 and combine 67 and 69?
fsync() takes the file and pushes pending writes...
As a system call, fsync() is expensive....